content/blog/thanos-monitoring.en.md

```diff
@@ -0,0 +1,524 @@
+---
+Title: Multi Cluster Monitoring with Thanos
+Date: 2021-02-23T11:00:00+00:00
+Category: Kubernetes
+Summary: Monitor multiple Kubernetes Cluster with Prometheus and Thanos
+Author: Kevin Lefevre
+image: images/logos/thanos.svg
+lang: en
+---
+
+### Introduction
+
+In this article we are going to see the limitation of a Prometheus only
+monitoring stack and why moving to a Thanos based stack can improve metrics
+retention and also reduce overall infrastructure cost.
+
+Content used for this demo is available
+[here](https://github.com/particuleio/teks/tree/main/terragrunt/live/thanos) and
+[there](https://github.com/particuleio/terraform-kubernetes-addons/tree/main/modules/aws)
+and are submit to their respective licenses.
+
+### Kubernetes Prometheus Stack
+
+When deploying Kubernetes infrastructure for our customer, it is standard to
+deploy a monitoring stack on each cluster. This stack is often comprise of
+several components:
+
+* [Prometheus](https://prometheus.io/): collect metrics
+* [AlertManager](https://prometheus.io/docs/alerting/latest/alertmanager/): send alerts to various provider based on metrics query
+* [Grafana](https://grafana.com/): fancy dashboards
+
+The simplify architecture is the following:
+
+![](/images/prometheus-thanos/prometheus-excalidraw-no-background.png)
+
+#### Caveats
+
+This architecture has some caveats and does not scale out well when increasing
+the number of cluster from which you want to get metrics.
+
+##### Multiple Grafana
+
+In this setup each cluster has its own Grafana with its own set of dashboards
+which can be a pain to maintain.
+
+##### Storing metrics is expensive
+
+Prometheus stores metrics on disk, you have to make a choices between storage
+space and metric retention time. If you want to store data for a long time and
+running on a Cloud Provider, block storage can be expensive if you store
+terabyte of data on it. Also in production environnement prometheus is often run
+either with replication or sharding or both, this can double or even quadruple
+your storage needs.
+
+#### Solutions
+
+##### Multiple Grafana Datasource
+
+It is possible to expose prometheus endpoints on the external network and to add
+them as Datasource in a single Grafana. You just need to implements security on
+the prometheus external endpoints with mutual TLS or TLS and basic auth for
+example. The drawback of this solution is that you cannot make calculation based
+on different data sources.
+
+##### Prometheus Federation
+
+[Prometheus
+federation](https://prometheus.io/docs/prometheus/latest/federation/) allow
+scraping prometheuses from prometheus, this solution works well when you are not
+scraping a lot of metrics. At scale if scraping all your other prometheuses
+takes longer than the scrape interval this is going to cause issue.
```

Comment:
What do you think of: `At scale, if the scrape duration of all your prometheus targets takes longer than the scrape interval, you might encounter some severe issues.` ?